Some docs improvements (#10426)
RasaHQ/rasa: docs/docs/migration-guide.mdx

`required_slots` method, note that `slots_mapped_in_domain` argument has been replaced by the `domain_slots` argument.
You will need to make this replacement to continue using your custom code.

:::note

> If you have custom validation actions extending `FormValidationAction` which override `required_slots` method, it is highly
> recommended you double-check the dynamic form behavior of your migrated assistant, in case slots set by the default action
> [`action_extract_slots`](./default-actions.mdx#action_extract_slots) might need to be reset within the context of your form by the custom validation actions.

Reviewer: ```suggestion
If you have custom validation actions extending `FormValidationAction` which override `required_slots` method, you should
double-check the dynamic form behavior of your migrated assistant. Slots set by the default action
[`action_extract_slots`](./default-actions.mdx#action_extract_slots) may need to be reset within the context of your form by the custom validation actions.
```

While sometimes there's no other way "it is recommended" should be avoided, especially when the recommendation is more like a global instruction

Reviewer: >Slots set by the default action
[`action_extract_slots`](./default-actions.mdx#action_extract_slots) may need to be reset within the context of your form by the custom validation actions.

How does this work? does the `FormValidationAction` do the resetting?

Does adding a short example make sense?

Author: Yup, the best example is the one created by Vincent that lead to this request in docs improvements.
For example, we have this domain:
```
forms:
  name_form:
    required_slots:
      - first_name
      - last_name
      
slots:
  first_name:
    type: text
    influence_conversation: true
    mappings:
      - type: from_text
        conditions:
         - active_loop: name_form
           requested_slot: first_name
  last_name:
    type: text
    influence_conversation: true
    mappings:
      - type: from_text
        conditions:
         - active_loop: name_form
           requested_slot: last_name
  name_spelled_correctly:
    type: bool
    influence_conversation: true
    mappings:
    - type: from_intent
      value: true
      intent: affirm
    - type: from_intent
      value: false
      intent: deny
```
`name_spelled_correctly` is a dynamic slot added via `FormValidationAction` and there's also a `utter_ask_name_spelled_correctly` with buttons.
But considering it uses a global slot mapping, when the user utters a message with the first name, that message gets misclassified as either `affirm` or `deny` and `action_extract_slots` will set the slot `name_spelled_correctly`. So what we want to do is that when `first_name` gets validated we want to reset the slot `name_spelled_correctly` to `None`, then this slot can be requested next.

```
    def validate_first_name(
        self,
        slot_value: Any,
        dispatcher: CollectingDispatcher,
        tracker: Tracker,
        domain: DomainDict,
    ) -> Dict[Text, Any]:
        """Validate `first_name` value."""
        # If the name is super short, it might be wrong.
        print(f"First name given = {slot_value} length = {len(slot_value)}")
        if len(slot_value) <= 1:
            dispatcher.utter_message(text=f"That's a very short name. I'm assuming you mis-spelled.")
            return {"first_name": None, "name_spelled_correctly": None}
        else:
            return {"first_name": slot_value, "name_spelled_correctly": None}
```
How best to add this example?

Reviewer: Tbh this example is still confusing to me because it's showing the reset in the context of validating a _different_ slot  - would need the dynamic addinig method too to make a complete example, but maybe that gets too long for this section. 
Is there a concise way to show in which method the resetting should happen, or is this entirely case dependant?

Author: > Is there a concise way to show in which method the resetting should happen, or is this entirely case dependant?

Depends case by case I'm afraid, as to when the dynamic slot is expected to be requested, is it after the first slot, 2nd slot in the form `required_slots` etc.

Reviewer: Gotcha. This is my attempt at a verbal example:
```
Slots set by the default action
[`action_extract_slots`](./default-actions.mdx#action_extract_slots) may need to be reset within the context of your form by custom validation actions for the form's required slots. For example, if your form dynamically adds a required slot after the first slot is filled, you may want to reset the potential required slot  as part of the first required slot's validation method to ensure it will be empty when added. 
```
Pretty confusing still but maybe better than nothing

Reviewer: Isn't it possible to overwrite a different method of the FormValidationAction to do this resetting based on the value of `requested_slot` itself?

Author: If you mean allow the developer overwrite a different method, there's nothing eligible from what's available in the interface. Even more, from what I can see [here](https://github.com/RasaHQ/rasa-sdk/blob/231e6a683cba9f7bc4e81f4972c46647f10a1f13/rasa_sdk/forms.py#L326), `next_requested_slot` method requires the dynamic slot to be set to `None` in order for it to be requested. But that's an interesting proposal, I'll submit a short issue in the sdk repo and drop it in our inbox, then prioritise according to community feedback.

Reviewer: That sounds promising! Let's merge the warning here as is then (generic warning to check behaviour), and then if this is a good solution we add it later.
